torch/package/package_exporter.py

```diff
@@ -585,24 +586,42 @@ def save_pickle(
             is_pickle=True,
         )
 
+        def _check_mocked_error(module: str, field: str):
+            for pattern, pattern_info in self.patterns.items():
+                if pattern_info.action == _ModuleProviderAction.MOCK:
+                    if pattern.matches(module):
+                        raise NotImplementedError(
+                            f"Object '{field}' from module {module} was mocked out during packaging "
+                            f"but is being used in resource - {resource} in package {package}"
+                            "If this error is happening during 'save_pickle', please ensure that your "
+                            "pickled object doesn't contain any mocked objects."
+                        )
+
         if dependencies:
             all_dependencies = []
+            module = None
+            field = None
             for opcode, arg, pos in pickletools.genops(data_value):
-                if opcode.name == "GLOBAL":  # a global reference
+                if sys.version_info >= (3, 4):
+                    if opcode.name == "SHORT_BINUNICODE" or opcode.name == "BINUNICODE8":
+                        assert isinstance(arg, str)
+                        module = field
+                        field = arg
+                    elif opcode.name == "BINGET_LONG" or opcode.name == "BINGET" or opcode.name == "GET":
+                        module = field
+                        field = None
+                    elif opcode.name == "STACK_GLOBAL":
+                        if module is None:
```

Comment:
Why would module ever be None? Isn't the point of STACK_GLOBAL to load a module?

Reply:
So this is specific to the set of 3 opcodes here. They grab some cached value which was likely already a global. Therefore, I just use None as a placeholder because we don't need to add the dependency again.

Reply:
After taking a closer look at https://github.com/python/cpython/blob/main/Lib/pickle.py, I think this would work the vast majority of the time. However, it isn't obvious. to me how, but a valid order of opcodes could be to memoize the module name and then call with w/o a prior save global. Which this doesn't cover. I'll keep a dict for the the memoization in case this is an issue.